internal/envoy/versions.go

```diff
@@ -20,30 +20,122 @@ import (
+	return &funcEVersions{ctx: ctx, envoyVersionsURL: envoyVersionsURL, platform: p, version: v}
+}
+
+type funcEVersions struct {
+	ctx              context.Context
+	envoyVersionsURL string
+	platform         version.Platform
+	version          version.Version
+
+	// Allow to override Get implementation.
+	getFunc funcEVersionsGetter
+}
+
+// Get returns a version map from a remote URL. e.g. from globals.DefaultEnvoyVersionsURL.
+func (f funcEVersions) Get() (version.ReleaseVersions, error) {
 	result := version.ReleaseVersions{}
 	// #nosec => This is by design, users can call out to wherever they like!
-	resp, err := httpGet(ctx, envoyVersionsURL, p, v)
+	resp, err := httpGet(f.ctx, f.envoyVersionsURL, f.platform, f.version)
 	if err != nil {
 		return result, err
 	}
 	defer resp.Body.Close() //nolint
 
 	if resp.StatusCode != http.StatusOK {
-		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, envoyVersionsURL)
+		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, f.envoyVersionsURL)
 	}
 	body, err := io.ReadAll(resp.Body) // fully read the response
 	if err != nil {
-		return result, fmt.Errorf("error reading %s: %w", envoyVersionsURL, err)
+		return result, fmt.Errorf("error reading %s: %w", f.envoyVersionsURL, err)
 	}
 
 	if err := json.Unmarshal(body, &result); err != nil {
 		return result, fmt.Errorf("error unmarshalling Envoy versions: %w", err)
 	}
+
 	return result, nil
 }
+
+// FindLatestVersions returns the latest patch version given an Envoy version.
+func (f funcEVersions) FindLatestVersions(envoyVersion version.Version) (string, error) {
+	if matched := globals.EnvoyVersionPattern.MatchString(string(envoyVersion)); matched {
+		return string(envoyVersion), nil
+	}
+
+	var latestPatch int
+	var latestVersion string
+
+	releases, err := f.versionsGetter()()
+	if err != nil {
+		return "", err
+	}
+
+	for v := range releases.Versions {
+		// The "." suffix is required to avoild false-matching, e.g. 1.1 to 1.18.
+		if !strings.HasPrefix(string(v), string(envoyVersion)+".") {
+			continue
+		}
+
+		var matched [][]string
+		if matched = globals.EnvoyMinorVersionPattern.FindAllStringSubmatch(string(v), -1); matched == nil {
+			continue
+		}
+
+		for _, sub := range matched {
+			// A matched patch component should look like ".4".
+			if !strings.HasPrefix(sub[1], ".") {
+				continue
+			}
+
+			var p int
+			if p, err = strconv.Atoi(sub[1][1:]); err != nil {
+				continue
+			}
+
+			if p >= latestPatch {
+				latestPatch = p
+				latestVersion = string(v)
+			}
+		}
+	}
+
+	return latestVersion, checkLatestVersion(latestVersion, string(envoyVersion))
+}
+
+// versionsGetter ensures the getter function is specified.
+func (f funcEVersions) versionsGetter() funcEVersionsGetter {
+	if f.getFunc != nil {
+		return f.getFunc
+	}
+	return f.Get
+}
+
+func checkLatestVersion(latest, v string) error {
```

Comment:
this isn't worth scrolling around for, can you inline the error check instead of making another function then having to name it and scroll to it?